### PR TITLE
Add a function which about check multiple gpu communicate by p2p

### PR DESCRIPTION
**Pull request by JunjieChen-2020**
Add a function which about check multiple gpu communicate by peer to peer available in single machine. And add a file which can test multiple gpus support p2p or not.

Reason:
nvidia is not support peer to peer by PCIe, hence, multiple gpus contact with each other by peer to peer need to fix NvLink. However, this hardware environment is not supported by every user, so tips are needed.

### CODE REVIEW

File: colossalai/communication/utils.py [hunk 1]
```diff
@@ -107,3 +107,34 @@ def gather_split_1d_tensor(tensor):
     chunks = [gathered[i*numel:(i+1)*numel] for i in range(world_size)]
     dist.all_gather(chunks, tensor, group=gpc.get_group(ParallelMode.PARALLEL_1D))
     return gathered
+
+import os
+import torch
+import warnings
+def check_single_machine_multi_gpu_p2p_available(gpu_list):
+    unavailble = []
+    check_p2p_result = os.popen("./simpleP2P")
```
Comment (feifeibear) [inline colossalai/communication/utils.py:116]: how dose this file generated?

File: colossalai/communication/utils.py [hunk 2]
```diff
@@ -107,3 +107,34 @@ def gather_split_1d_tensor(tensor):
...
+    p2p_info = check_p2p_result.read().split('\n')
+    for index in range(len(p2p_info)):
```
Comment (feifeibear) [inline colossalai/communication/utils.py:118]: use enumerate instead range

File: colossalai/communication/utils.py [hunk 3]
```diff
@@ -107,3 +107,34 @@ def gather_split_1d_tensor(tensor):
...
+        if "No" in p2p_info[index]:
+            unavailble.append(p2p_info[index])
+    print(unavailble)
```
Comment (feifeibear) [inline colossalai/communication/utils.py:121]: do not use print. And this line seems used for debugging.

File: colossalai/communication/utils.py [hunk 4]
```diff
@@ -107,3 +107,34 @@ def gather_split_1d_tensor(tensor):
...
+    count = 0
+    if len(unavailble) == 0:
+        return True
+    elif torch.cuda.device_count() >= len(unavailble):
+        for p2p_unavailable_gpu in unavailble:
+            gpu_id = p2p_unavailable_gpu.split(" ")[6][4:5]
+            print(gpu_id)
```
Comment (feifeibear) [inline colossalai/communication/utils.py:128]: debugging?

File: colossalai/communication/utils.py [hunk 5]
```diff
@@ -107,3 +107,34 @@ def gather_split_1d_tensor(tensor):
...
+            if gpu_id in gpu_list:
+                warnings.warn(p2p_unavailable_gpu)
```
Comment (feifeibear) [inline colossalai/communication/utils.py:130]: show more warning info please

File: colossalai/communication/utils.py (hunk 1 above)
Comment (JunjieChen-2020) [reply colossalai/communication/utils.py:116]: The file located in /usr/local/cuda-10.1/samples/0_Simple/simpleP2P.

Comment (JunjieChen-2020) [reply colossalai/communication/utils.py:116]: We can use simpleP2P to test peer to peer after makefile.

File: colossalai/communication/utils.py (hunk 2 above)
Comment (JunjieChen-2020) [reply colossalai/communication/utils.py:118]: ok

File: colossalai/communication/utils.py (hunk 3 above)
Comment (JunjieChen-2020) [reply colossalai/communication/utils.py:121]: ok

File: colossalai/communication/utils.py (hunk 4 above)
Comment (JunjieChen-2020) [reply colossalai/communication/utils.py]: Select the information of check: whether the gpus in the parameter gpu_list submitted by the user can P2P

File: colossalai/communication/utils.py [hunk 6]
```diff
@@ -107,3 +108,43 @@ def gather_split_1d_tensor(tensor):
     chunks = [gathered[i*numel:(i+1)*numel] for i in range(world_size)]
     dist.all_gather(chunks, tensor, group=gpc.get_group(ParallelMode.PARALLEL_1D))
     return gathered
+
+
+def check_single_machine_multi_gpu_p2p_available(gpu_list):
+    """Opposite of above function, gather values from model parallel ranks.
```
Comment (feifeibear) [inline colossalai/communication/utils.py:114]: I guess you wrongly copy the comments and forget to chang it.

File: colossalai/communication/utils.py [hunk 7]
```diff
@@ -107,3 +108,43 @@ def gather_split_1d_tensor(tensor):
...
+
+    :param gpu_list: Queue of gpu to be tested
```
Comment (feifeibear) [inline colossalai/communication/utils.py:116]: may be a list of gpu more appropriate

File: colossalai/communication/utils.py [hunk 8]
```diff
@@ -107,3 +108,43 @@ def gather_split_1d_tensor(tensor):
...
+    :type list
+
+    :return True or False: gpus returns True if both can access each other via peer to peer, otherwise returns False
+    :rtype bool
+    """
+    logging.basicConfig(format='%(asctime)s - %(pathname)s[line:%(lineno)d] - %(levelname)s: %(message)s', level = logging.DEBUG)
+    unavailble = []
+    check_p2p_result = subprocess.Popen("./simpleP2P", shell=False, stdout=subprocess.PIPE)
```
Comment (feifeibear) [inline colossalai/communication/utils.py:124]: I do not think it's a good idea to pack a 700KB binary file inside the source code.

**Comment by feifeibear** [comment]
Thanks for your contribution. Please explain why we need the ```check_single_machine_multi_gpu_p2p_available```function?

File: colossalai/communication/utils.py (hunk 8 above)
Comment (JunjieChen-2020) [reply colossalai/communication/utils.py:124]: python may not be able to check if peer to peer is available.

**Review by JunjieChen-2020 — COMMENTED** [review]
Modified some description statements.

**Pull request closed without merging**